src/ShipAICmd.cpp

```diff
@@ -722,8 +723,8 @@ static vector3d GenerateTangent(DynamicBody *dBody, FrameId targframeId, const v
 //4 - probable path intercept
 int CheckCollision(DynamicBody *dBody, const vector3d &pathdir, double pathdist, const vector3d &tpos, double endvel, double r)
 {
-	if (!dBody->Have(DynamicBody::PROPULSION)) return 0;
-	Propulsion *prop = dBody->GetPropulsion();
+	if (!dBody->HasComponent<Propulsion>()) return 0;
```

Comment:
It seems that in about all cases it will be `true`, which means we will almost always do the same thing twice. Maybe just return 0 if `GetComponent` returned `nullptr`?

Reply:
Because the underlying container is a `std::map`, GetComponent will throw an exception if the body does not have the specified component. HasComponent is pretty fast, as it's a single hashtable lookup and a bitset index.

I may eventually move to a model where GetComponent returns `nullptr` if the underlying datastructure is amenable to it.

Reply:
but no, `GetComponent` does a check identical to `HasComponent` before accessing.

Reply:
You are indeed correct, my apologies. It's been so long since I initially wrote this all the little details have slipped from my mind.